### PR TITLE
CUPS migration actors

### PR DESCRIPTION
**Pull request by zdohnal**
Hi!

I created two actors for cups migration - cupscheck and cupsmigrate. The code is currently 'working prototype', tested only manually for now.
I will move all 'non-actor' methods to outside module and create unit tests after my PTO, would you mind reviewing the code in meanwhile?

Thank you in advance!

Zdenek

### CODE REVIEW

File: repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py]: Could we have a more descriptive name for this model? Something like CupsMigrationMatrix? That way we can avoid some naming conflicts in the future. Also, some docstring documentation about the model can be helpful.

File: repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py]: ditto about having docstring documentation.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: cupsd_changes_list could be a Python's dictionary (or even namedtuples) what would simplify this code a bit. Instead of:
```python
INCLUDE = 0
(...)
cupsd_changes_list [['include', (...)]]
(...)
cupsd_changes_list[INCLUDE]
```
we can have:

```python
cupsd_changeslist {'include': ['include', (...)]}
cupsd_changes['include']
```
Also, since this is used at a single place afaict, moving it closer to where it is used (inside function) can help readability in my opinion.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: ditto for using a dictionary or namedtuples.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Would be interesting to return the error attached to the IOError exception. This can help us or users to understand better why an error happened.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: ditto for returning a descriptive error.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Same thing for this function, instead of returning boolean maybe we could return a descriptive error if something happened or empty string if everything is OK. Also, maybe would be a good practice to check if all files exists and are readable/writeable before doing proper changes. If the process fails in the end, but some changes were applied we can end up with a not working system.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: ditto about returning meaningful errors and doing all checks before properly executing something.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: In this function there are some uses of `if variable is True:` and `if variable is False:`. We could use `if variable:` and `if not variable:` instead.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: If no change was made and the user should not take any action, I would advocate that we should not add an entry to the report. In that case I would remove this use of report_generic.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: No need to raise an exception here, we could just use `return`.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: My other comments about functions returning meaningful error messages is related with this raised exception (and following ones). This error will stop Leapp processing (not only this actor), if that is what we want we should make it clear why it failed and how to solve the problem.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: If changes on CUPS are now transparent to user and the system will be working exactly it was before, I would remove this report_generic. If the changes made can have some impact into daily use of system I would add some detailed message about what was changed.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: As mentioned on other file, I have the feeling that this can be better represented using a dictionary or namedtuples.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: It is not necessary to close() opened_config. After the code exits current `with` context, the opened file will be closed automatically.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: ditto for no need to close() opened_config.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: No need for `is True` or `is False` at if statements.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Since no action is needed, I would remove this report entry.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Since no action is needed, I would remove this report entry.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (artmello) [inline repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: The process will happen (and succeed or fail) without user intervention. I think we can remove this entry from report and only notify user that migration was made.

**Comment by zdohnal** [comment]
Hi everyone,

I'm still working on the actor for now, I'll let you know here in the ticket when it will be ready for new review.

**Comment by bocekm** [comment]
Hi @zdohnal. Would you please try finishing this PR. Thanks.

File: repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py]: No longer used.

File: repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py]: Docstring added.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: I chose different approach in the end.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Using different approach now.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Done.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Noted, fixed.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Corrected.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Removed.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Removed.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Removed.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Corrected.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Now I check cupsd.conf and cups-files.conf for availability before doing anything else, exiting with error message if CUPS is installed but confs are missing.
Ad included files - their absence does not crash CUPS, just adds error into log - IMO it is user error to put non-existing file into configuration. It generates error (but service is still running) even on RHEL 7, so I chose to let 'Include' directives with non-existing files in cupsd.conf.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Done.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Noted. Fixed.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: No longer creates reports in this actor.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Ok.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: Removed.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py]: No longer creates report in this actor.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Used different approach, hopefully more understandable.

File: repos/system_upgrade/el7toel8/actors/cupscheck/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/actor.py]: Noted. Fixed.

**Comment by zdohnal** [comment]
Ready for review after refactoring. Thank you for your patience!

**Comment by pirat89** [comment]
Thanks :)

**Comment by pirat89** [comment]
e2e tests

**Comment by pirat89** [comment]
@zdohnal Let's sync on monday. I think it is ok, but regarding UX we will need some changes in reports here. I can do all changes for you if you are run out of time regarding the point we missed that in previous reviews, the code seems clear to me.

**Comment by zdohnal** [comment]
@pirat89 Ok, lets sync today or this week.

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: keep separate

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: conf group 1. (severity none)

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: conf group 2.

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: conf group 2.

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: conf group 1. but:
- in case the cert and keys are in different directories, they will be copied into the `/etc/cups/ssl` directory (otherwise no action is needed). In such case, after change of keys or certs, both have to be updated, or what is recommended, to check your setup, modify scripts if needed to use the new directory and remove original ones maually. (msg draft)

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py]: note for me: conf group 1.

**Comment by AloisMahdal** [comment]
Note about tests: The `static` result is false positive (infra issue); the actual test passed and in a while (once I internally release morf-0.5.1) the part that caused problems will be disabled so you can then re-run tests.

OTOH, the `upgrade` part is valid; the upgrade went to emergency mode which led to timeout after total of 8h.  I spoke to @pirat89 about it; I'm assuming he's investigating it. (Please correct me if I'm wrong..)

**Comment by pirat89** [comment]
@AloisMahdal: You are right.

**Comment by drehak** [comment]
leapp-ci build

**Comment by AloisMahdal** [comment]
FTR, the failing tests above are probably (logs are garbage-collected already) for infra reasons that should be mostly fixed now.

**Comment by bocekm** [comment]
@zdohnal, what is your plan with this PR? Is it WIP or do you need any addition review/help from the leapp team?

**Comment by pirat89** [comment]
@bocekm that's waiting for our investigation why the upgrade has been crashing.

**Comment by zdohnal** [comment]
Hi @pirat89 ,

can I help anyhow at the moment?

**Comment by AloisMahdal** [comment]
Can't say what to do with open peer review above, but there's been some progress on master in the meantime so rebasing & re-running `e2e tests` could be a good start...

**Comment by AloisMahdal** [comment]
> @bocekm that's waiting for our investigation why the upgrade has been crashing.

@pirat89 not sure which crash you meant but if it was in beaker then logs are looong time gone.

**Comment by zdohnal** [comment]
@AloisMahdal  @pirat89 Tests failed after rebasing to the current master with import errors like:

```
_ ERROR collecting repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/tests/test_digest_directive_check.py _

ImportError while importing test module '/payload/repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/tests/test_digest_directive_check.py'.

Hint: make sure your test modules/packages have valid Python names.

Traceback:

/payload/tut/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule

    mod = self.fspath.pyimport(ensuresyspath=importmode)

/payload/tut/lib/python2.7/site-packages/py/_path/local.py:704: in pyimport

    __import__(modname)

/payload/tut/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:304: in load_module

    exec(co, mod.__dict__)

tests/test_digest_directive_check.py:3: in <module>

    from leapp.libraries.actor.library import digest_directive_check

E   ImportError: cannot import name digest_directive_check
```

Those errors didn't happen in previous testing. I checked for the error on the net and found a piece of information that it can be caused by __init__.py adding/removal, but no other actors I checked - quagga, bind, vsftpd - have it...
Would you mind helping me with it?

**Comment by pirat89** [comment]
@zdohnal Definitely. We will sync with you in the next two weeks. Just finishing works around the next release.

@AloisMahdal yes, it was in beaker.

**Comment by zdohnal** [comment]
Hi @pirat89 ,

can we sync about the PR next week and figure out the plan what to do for merging it?

Thank you in advance!

Zdenek

**Comment by pirat89** [comment]
@zdohnal Hi, I am sorry I have forgotten about that completely. We had two very close releases recently. Yes, I will have definitely a time during the next week to sync with you. I have scheduled the mtg in our calendar.

**Comment by pirat89** [comment]
Discussed, we will go to push the work around Nov/Sep.

File: repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:44]: @pirat89 my guess - the idea about 'groups' could be a aggregation of reports based on severity:

0) 'Drivers' report will stay separate
1) all other 'HIGH' reports will be merged in 1 func
2) all 'MEDIUM' reports will be merged together in 1 func

Let me know after you have time for looking into this.

File: repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/actor.py
Comment (zdohnal) [reply repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/actor.py]: Renamed in the newest version.

**Comment by zdohnal** [comment]
Ok, I updated the PR, fixed the actor based on @pirat89 findings (actor's libraries mustn't have a generic name, otherwise the import will fail) and incorporated two of suggested changes - renaming cupsfeaturesfind to cupsscanner and using StopActorExecution error if conf files are missing.

@pirat89 then the only thing remaining is a conf groups, which I don't recall exactly what supposed to mean - my guess is an aggregation of reports either by severity, tag or number of affected files, but I'm not sure. If you manage to check it earlier and write down the expectations, I can try to update the code accordingly.

Ad failed e2e test: @AloisMahdal would you mind helping me with e2e test [results](https://baseos-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/oamg-ci-test-internal_copr-ts_sti-artemis/213/artifact/consoleText.txt.gz)? I see there a host cannot be reached, is there something I can do?

**Comment by mkluson** [comment]
e2e test fail: the machine did not boot after the upgrade, I will rerun it once again with different parameters to get serial console log

**Comment by AloisMahdal** [comment]
Still failing.. Not sure why.  I'm running Beaker-based Murphy suite as backup or for comparison.

**Comment by AloisMahdal** [comment]
Murphy upgrade test got stuck too, because there's traceback & emergency mode:

        [..]
        [..]
    [  784.597300] upgrade[593]: 2020-11-10 08:39:05.443 DEBUG    PID: 839 leapp.fallback: Updating Vim configuration file /etc/virc. 
    [  784.647614] upgrade[593]: 2020-11-10 08:39:05.495 INFO     PID: 3 leapp.workflow.Applications: Executing actor migrate_brltty 
    [  784.705840] upgrade[593]: 2020-11-10 08:39:05.553 INFO     PID: 3 leapp.workflow.Applications: Executing actor cupsfilters_migrate 
    [  784.801243] upgrade[593]: 2020-11-10 08:39:05.648 INFO     PID: 3 leapp.workflow.Applications: Executing actor cups_migrate 
    [  784.854298] upgrade[593]: Process Process-183: 
    [  784.855407] upgrade[593]: Traceback (most recent call last): 
    [  784.856554] upgrade[593]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap 
    [  784.858127] upgrade[593]:     self.run() 
    [  784.859028] upgrade[593]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run 
    [  784.860519] upgrade[593]:     self._target(*self._args, **self._kwargs) 
    [  784.861731] upgrade[593]:   File "/root/tmp_leapp_py3/leapp/repository/actor_definition.py", line 72, in _do_run 
    [  784.863337] upgrade[593]:     actor_instance.run(*args, **kwargs) 
    [  784.864450] upgrade[593]:   File "/root/tmp_leapp_py3/leapp/actors/__init__.py", line 335, in run 
    [  784.865996] upgrade[593]:     self.process(*args) 
    [  784.868378] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/actor.py", line 21, in process 
    [  784.870180] upgrade[593]:     cupsmigrate.migrate_configuration() 
    [  784.871326] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py", line 180, in migrate_configuration 
    [  784.873383] upgrade[593]:     facts = consume_function(CupsChangedFeatures) 
    [  784.874685] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py", line 164, in _get_facts 
    [  784.876685] upgrade[593]:     return next(api.consume(model)) 
    [  784.877813] upgrade[593]: StopIteration 
    [  784.889279] upgrade[593]: ============================================================================================= 
    [  784.891192] upgrade[593]: Actor cups_migrate unexpectedly terminated with exit code: 1 - Please check the above details 
    [  784.892879] upgrade[593]: ============================================================================================= 
        [..]
        [..]


That was with:

 *  *leapp-0.11.1-100.202011021301Z.90b67be.master.el7_6*
 *  *leapp-repository-0.12.0-0.202011100708Z.950cddf.cups_migration.PR181.el7_6*

**Comment by AloisMahdal** [comment]
Note that the above was clean installation, ie. cups was not installed.

**Comment by zdohnal** [comment]
Aha, the default value was missing from ```next()``` call in ```cupsmigrate``` actor - ```None``` :). Fixed now and the testsuite was updated to check this case.

**Comment by Rezney** [comment]
@zdohnal @pirat89 this PR seems to be "a little" stuck. Would you have time to get it back on track if it is still valid?

**Comment by pirat89** [comment]
I will look at it today/tomorrow.

File: repos/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py:230]: @zdohnal we should ad here `TODO: add post upgrade report`. As we are not doing post-upgrade reports so much yet, we do not have to add it now. But we should add least add TODO here.

**Comment by pirat89** [comment]
@zdohnal so far it seems ok and working. However I am still confused about the latest change "startswith" -> "rsplit". Probably I am working too long today already. Let's sync tomottor. I've tested the solution for `PrintcapFormat` and everything has been done as expected.

And I want to discuss the TLS one more time in cupsd (I mean, `Digest` -> `Basic` change).

**Comment by zdohnal** [comment]
> @zdohnal so far it seems ok and working. However I am still confused about the latest change "startswith" -> "rsplit".

To sum it up, "startswith" is 'vulnerable' to substrings - f.e.:

- server.key lives in /etc/cups/ssl
- server.cert lives in /etc/cups/ssl/smdir

'startswith', if the former gets into certkey_values first, then the latter matches as substring and the function incorrectly puts ServerKeychain /etc/cups/ssl and no copying happens -> cupsd won't be using the correct cert.

If we stayed with 'startswith', we would need to 'preprocessed' certkey_values or choose the value with the longest path instead of the first value in the list. In that case 'rsplit' function looked cleaner to get the filename and its path, and than do the exact comparison.

> And I want to discuss the TLS one more time in cupsd (I mean, `Digest` -> `Basic` change).

Ok, basically how the users do the remote auth doesn't change - the thing which changes is how 'protected' are creds - with Digest, the creds were hashed by hashing alg and sent over unencrypted connection. with Basic, the creds are in plain text, but the connection is protected by TLS, thus not that easy to leak.

**Review by pirat89 — APPROVED** [review]
It's here!!! Waiting for QE approval!

**Review by pirat89 — APPROVED** [review]
just regular rebase, everything is same as before. re-ack

**Comment by pirat89** [comment]
btw, in case the e2e fails, ignore it. the test is broken nowadays and it's going to be removed soon.

**Comment by zdohnal** [comment]
\o/ thx for merging it :)

**Comment by pirat89** [comment]
Thanks for all your hard work you did in this PR!!

**Pull request merged**
